pkg/controller/storagecluster/kms_resources.go

```diff
@@ -0,0 +1,56 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	ocsv1 "github.com/openshift/ocs-operator/pkg/apis/ocs/v1"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *ReconcileStorageCluster) ensureKMSSystem(instance *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	// if there is no 'Security' spec to inspect, return silently
+	if instance.Spec.Security == nil {
+		return nil
+	}
+	if instance.Spec.Security.KeyManagementService != nil &&
+		instance.Spec.Security.KeyManagementService.ConnectionDetails == "" {
+		// if there is 'ConnectionDetails' configmap name is not provided, then return silently
+		return nil
+	}
+	ctx := context.TODO()
+	// Check for the CephCluster
+	found := &cephv1.CephCluster{}
+	err := r.client.Get(ctx,
+		types.NamespacedName{
+			Name:      generateNameForCephCluster(instance),
+			Namespace: instance.Namespace,
+		},
+		found,
+	)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("CephCluster yet to start."))
+		return err
+	}
+	kmsConfigMapName := strings.TrimSpace(instance.Spec.Security.KeyManagementService.ConnectionDetails)
+	kmsConfigMap := &corev1.ConfigMap{}
+	err = r.client.Get(ctx,
+		types.NamespacedName{
+			Name:      kmsConfigMapName,
+			Namespace: instance.ObjectMeta.Namespace,
+		},
+		kmsConfigMap,
+	)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Unable to get '%s' config map", kmsConfigMapName))
+		return err
+	}
+	found.Spec.Security.KeyManagementService.ConnectionDetails = kmsConfigMap.Data
```

Comment:
Why not do this at the time the CephCluster object is created?Is there a reason this is being done separately?

Reply:
Agreed, let's just build the encryption config altogether with the other CR options.

Reply:
Yes. This can be added to newCephCluster(). No need for another ensure function if no new CR is being created.

Reply:
Yup I agree and made the change...

There were questions around here like,
a. What if we change the data in config map (like VAULT_ADDR, PORT etc), we are not watching the objects?
b. What happens when token or config map get deleted?